skrub/_table_vectorizer.py

```diff
@@ -495,147 +435,101 @@ def _clone_transformers(self):
-                                )
-                    except (ValueError, TypeError):
-                        pass
-            # Cast pandas dtypes to numpy dtypes
-            # for earlier versions of sklearn. FIXME: which ?
-            if issubclass(X[col].dtype.__class__, ExtensionDtype):
-                try:
-                    X[col] = X[col].astype(X[col].dtype.type, errors="ignore")
-                except (TypeError, ValueError):
-                    pass
-            self.types_[col_idx] = X[col].dtype
-        return X
+            # Cast pandas dtypes to numpy dtypes for earlier versions of sklearn.
+            # Categorical dtypes don't need to be casted.
+            # Note that 'is_category_dtype' is deprecated.
+            if is_extension_array_dtype(X[col]) and not isinstance(
+                X[col].dtype, CategoricalDtype
+            ):
+                dtype = X[col].dtype.type
+                X[col] = X[col].astype(dtype)
 
-    def _apply_cast(self, X: pd.DataFrame) -> pd.DataFrame:
-        """Takes a dataframe, and applies the best data types learnt during fitting.
+                # When converting string to object, <NA> values becomes '<NA>'
+                # so we need to replace false missing values once more.
+                X[col] = _replace_false_missing(X[col])
+
+            # For object dtype columns, convert to string to avoid mixed types.
+            if is_object_dtype(X[col]):
+                X.loc[X[col].notnull(), col] = X[col].astype(str)
+
+        if reset:
+            X = to_datetime(X)
+            X = _to_numeric(X)
+            self.type_per_column_ = X.dtypes.to_dict()
+
+        else:
+            # Update categorical dtype.
+            error_msg = (
+                "This %(name)s instance is not fitted yet. "
+                "Please run auto_cast(X, reset=True) first."
+            )
+            check_is_fitted(self, ["type_per_column_"], msg=error_msg)
+            category_columns = X.select_dtypes("category").columns
+            for col in category_columns:
+                dtype = self.type_per_column_[col]
+                dtype = _union_category(X[col], dtype)
+                self.type_per_column_[col] = dtype
+                X[col] = X[col].astype(dtype)
+
+            # Enforce dtypes conversion using the dtypes seen during fit.
+            # As this behavior is more aggressive than skrub's to_datetime
+            # or _to_numeric and only makes sense for the TableVectorizer, we
+            # define it here rather than within these two functions.
+            # See: https://github.com/skrub-data/skrub/issues/837
+            for column, dtype in self.type_per_column_.items():
+                if is_numeric_dtype(dtype):
+                    X[column] = pd.to_numeric(X[column], errors="coerce")
+
+                elif is_datetime64_any_dtype(dtype):
+                    X[column] = pd.to_datetime(X[column], errors="coerce")
```

Comment:
IIRC we wanted to fit a DatetimeEncoder during train and use it to parse dates (with the same format) in transform?

Reply:
Addressed in the follow-up PR: https://github.com/skrub-data/skrub/pull/848